### PR TITLE
[libc++] Increase `atomic_ref`'s required alignment for small types

### PR DESCRIPTION
**Pull request by dalg24**
Background discussion here https://github.com/llvm/llvm-project/pull/99570#issuecomment-2237668661

Require types that are 1, 2, 4, 8, or 16 bytes in length to be aligned to at least their size to allow them to be used lock-free.

cc @jyknight @ldionne

### CODE REVIEW

File: libcxx/include/__atomic/atomic_ref.h
Comment (jyknight) [inline libcxx/include/__atomic/atomic_ref.h:100]: typo: `alignement` should be `alignment`.

File: libcxx/include/__atomic/atomic_ref.h
Comment (jyknight) [inline libcxx/include/__atomic/atomic_ref.h:98]: I'd say "to be potentially used lock-free". We don't know if they _actually_ can or not on a given target.

Comment (dalg24) [reply libcxx/include/__atomic/atomic_ref.h]: Done.  Thanks.

**Comment by jyknight** [comment]
Thanks! LGTM, but wait for ldionne to approve.

File: libcxx/include/__atomic/atomic_ref.h
Comment (ldionne) [inline libcxx/include/__atomic/atomic_ref.h:100]:
```suggestion
  static constexpr size_t __min_alignment = (sizeof(_Tp) & (sizeof(_Tp) - 1)) || (sizeof(_Tp) > 16) ? 0 : sizeof(_Tp);
```

Did you really mean `bool`?

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/include/__atomic/atomic_ref.h
Comment (dalg24) [reply libcxx/include/__atomic/atomic_ref.h:100]: Nope, def size_t

**Review by ldionne — APPROVED** [review]
LGTM if CI is green. Whoever merges, please cherry-pick to `release/19.x`.

**Comment by ldionne** [comment]
I'd like to better understand the failure mode here when passing a variable that isn't suitably aligned. It seems that with this patch it becomes pretty easy to violate the `std::atomic_ref::required_alignment`, as can be seen by the failing tests. For example:

```
struct LLIArr2 { long long int i[2]; }
```

Just naively creating a `LLIArr2 x` on the stack and then using it inside an `atomic_ref` can either work or fail the `required_alignment` check depending on the address at which the variable landed. We only catch that loudly in the Extensive and Debug hardening modes -- otherwise it's silent.

What happens when the alignment requirement is not satisfied? Is it only that we will incorrectly report that the `atomic_ref` is always lock free when in reality it uses a lock under the hood due to the incorrect alignment?

**Comment by dalg24** [comment]
> What happens when the alignment requirement is not satisfied? Is it only that we will incorrectly report that the `atomic_ref` is always lock free when in reality it uses a lock under the hood due to the incorrect alignment?

This is also how I understand it.

**Comment by jyknight** [comment]
> What happens when the alignment requirement is not satisfied? Is it only that we will incorrectly report that the `atomic_ref` is always lock free when in reality it uses a lock under the hood due to the incorrect alignment?

On common platforms, you may get either silently-non-atomic accesses or a crash.

**Comment by jyknight** [comment]
Ugh...actually...lemme take that back -- we aren't using that alignment info, currently. So I am wrong about the current behavior, and you are right.

What I said is what we should _want_ to be the case, but it's not now.

This PR is still an improvement, and makes the ABI change that's required. But the purpose of requiring alignment is to generate efficient accesses, and this doesn't actually get there, yet. We instead generate a libcall (which, if the object is in fact aligned to required_alignment, will be lock-free), but it's still overhead of a should-be unnecessary libcall.

**Comment by jyknight** [comment]
OK, to fix it, I think we just need:
```
  typedef _Tp _Aligned_Tp __attribute__((aligned(required_alignment)))`, 
  _Aligned_Tp* __ptr_;
```
(moving decls around as needed)

File: libcxx/include/__atomic/atomic_ref.h
Comment (EricWF) [inline libcxx/include/__atomic/atomic_ref.h:105]: I think we should calculate the alignment using something like this?

```
template <class T>
constexpr size_t calc_alignment() {
    size_t align = alignof(T);
    while (align <= alignof(max_align_t)) {
        if (__atomic_always_lock_free(sizeof(T), align))
            return align;
        align *= 2;
    }
    return alignof(T);
}
```

Comment (ldionne) [reply libcxx/include/__atomic/atomic_ref.h:105]: I really like this, I think this directly translates to what we are actually trying to achieve (which is to calculate the smallest alignment that allows us to be lockfree).

Comment (jyknight) [reply libcxx/include/__atomic/atomic_ref.h:105]: That is not a valid implementation, because it causes the ABI of the type to change depending on the microarchitecture selection.

Comment (ldionne) [reply libcxx/include/__atomic/atomic_ref.h:105]: TIL

**Comment by ldionne** [comment]
@dalg24 Can you please rebase this on top of https://github.com/llvm/llvm-project/pull/99570 I just merged and also apply @jyknight 's comment to use `__attribute__((aligned(required_alignment)))`? This should allow the compiler to assume that the pointer is properly aligned and generate more optimal code.

**Review by ldionne — COMMENTED** [review]
This LGTM but now the tests are failing because we pass an insufficiently-aligned object as part of our tests. We might want to use `alignas(atomic_ref<T>::required_alignment)` on the objects we pass to `atomic_ref` in the test?

**Pull request merged**
